nixos/modules/services/development/livebook.nix

```diff
@@ -0,0 +1,75 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  cfg = config.services.livebook;
+in
+{
+  options.services.livebook = {
+    enableUserService = mkEnableOption "a user service for Livebook";
+
+    environmentFile = mkOption {
+      type = types.path;
+      description = lib.mdDoc ''
+        Environment file as defined in {manpage}`systemd.exec(5)` passed to the service.
+
+        This must contain at least `LIVEBOOK_PASSWORD`.  See `livebook server --help` for
+        other options.
+      '';
+    };
+
+    erlang_node_short_name = mkOption {
+      type = with types; nullOr str;
+      default = null;
+      example = "livebook";
+      description = "A short name for the distributed node.";
+    };
+
+    erlang_node_name = mkOption {
+      type = with types; nullOr str;
+      default = null;
+      example = "livebook@127.0.0.1";
+      description = "The name for the app distributed node.";
+    };
+
+    port = mkOption {
+      type = with types; nullOr port;
+      default = 8080;
+      description = "The port to start the web application on.";
+    };
+
+    cookie = mkOption {
+      type = with types; nullOr str;
+      default = null;
+      description = "The cookie for the app distributed node.";
+    };
+
+    address = mkOption {
+      type = with types; nullOr str;
+      default = "127.0.0.1";
+      description = lib.mdDoc ''
+        The address to start the web application on.  Must be a valid IPv4 or
+        IPv6 address.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enableUserService {
+    systemd.user.services.livebook = {
```

Comment:
what considerations did you go through to choose a user service rather than a system service ? was that just as a start ?

Reply:
A user service is my use-case.  I think I've written it such that a system service can be easily added later, but I don't personally have an interest in that.

Reply:
Youve done 90% of the hard work for a system service, and i dont mind pushing it over the finish line.
If its just a matter of convenience, after we merge this im happy to make an additional commit to switch this to a system service.
Do you have any objections with this being a system service instead of a user service ?
(I think it makes more sense )

Reply:
Maybe I object to it being exclusively a system service.  I don't understand how some things would work.

How would security work?  Wouldn't that just give all users with access to the notebooks root access?

What happens when the server saves a notebook as a *.livemd file?  Would that be owned by root?  That sounds like it would be annoying for users.

Reply:
A more livebook specific question: it's possible to connect a notebook to a running node.  You might want to do this if you want to use livebook to control the apps running on that node of if you want to access data on that node.  I think `epmd` will just enable this to work even if livebook and the node have different users, but I think this then gives all the users with access to livebook full access as the user running the attached node.

I guess what I'm saying with all these questions is that there are two ways to run livebook:

A. Each user runs livebook as their own tool and they don't care about security at all because they're the only ones with access.
B. Somebody runs livebook for multiple users.  They care about security, so they lock down the systemd service as much as possible.  They probably also then force livebook to only use S3 as storage or something else that's shared.

I think A basically has to be a user service and B has to be a system service.  I'm personally only interested in use-case A.

Reply:
- on the concept of who runs the system service, you can create a user "livebook" that will run the service, and set a group and the group access accordingly. So I don't think there is a problem of running as root.
- on the concept of authentication there are 3 providers used for authentication https://hexdocs.pm/livebook/authentication.html.

That said though, I agree with what you are saying as running a livebook as a single user and not caring about much (I think that will be the most common use case in nixos).
The fact that livebook doesn't have a granular system of permission makes a good argument for running this as a user service.
Maybe we should add a comment in the tune of 
"Since livebook doesn't have a granular permission system (a user either has access to all the data or none at all), the decision was made to run this as a user service, if that changes in the future, this can be changed to a system service".

Reply:
Ok.  I've added the comment to the Nix service file.